File: src/gsettings.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>

const char* FILENAME =
    "/usr/share/glib-2.0/schemas/mint-artwork.gschema.override";

int main(int argc, char** argv) {
  if (argc != 5) {
    std::cout << "usage: ./gsettings set SCHEMA KEY VALUE" << std::endl;
    return 1;
  }

  std::string schema = argv[2];
  std::string key = argv[3];
  std::string value = argv[4];

  schema = "[" + schema + "]";
  key += "=";

  std::ifstream infile(FILENAME);
  std::ostringstream sstr;
  sstr << infile.rdbuf();
  std::string buffer = sstr.str();

  size_t start = buffer.find(schema);
  if (start != std::string::npos) {
    size_t end = buffer.find("\n\n[", start);
    if (end == std::string::npos)
      end = buffer.length() - 1;
    size_t len = end - start;
    std::string subbuffer = buffer.substr(start, len);

    size_t substart = subbuffer.find(key);
    if (substart != std::string::npos) {
      substart += key.length();
      size_t subend = subbuffer.find("\n", substart);
      size_t sublen = subend - substart;
      subbuffer.replace(substart, sublen, value);
    } else {
      subbuffer += "\n" + key + value;
    }

    buffer.replace(start, len, subbuffer);
  } else {
    buffer += "\n" + schema;
    buffer += "\n" + key + value;
    buffer += "\n";
  }

  std::ofstream outfile(FILENAME);
  outfile << buffer;
}

File: src/gsettings.rs
use std::env;
use std::fs;

const PATH: &str = "/usr/share/glib-2.0/schemas/mint-artwork.gschema.override";

fn main() {
    let args: Vec<String> = env::args().collect();

    if args.len() < 5 {
        panic!("usage: ./gsettings set SCHEMA KEY VALUE");
    }

    let schema = format!("[{}]", args[2]);
    let key = format!("{}=", args[3]);
    let value = args[4].clone();

    let mut contents = fs::read_to_string(PATH).unwrap();
    if let Some(start) = contents.find(&schema) {
        let end = match contents[start..].find("\n\n[") {
            Some(idx) => start + idx,
            None => contents.len() - 1,
        };

        let mut substring = String::from(&contents[start..end]);
        if let Some(substart) = substring.find(&key) {
            let subend = match substring[substart..].find("\n") {
                Some(idx) => substart + idx,
                None => substring.len(),
            };

            substring.replace_range(substart..subend, &format!("{}{}", key, value));
        } else {
            substring += &format!("\n{}{}", key, value);
        }

        contents.replace_range(start..end, &substring);
    } else {
        contents += &format!("\n{}\n{}{}\n", schema, key, value);
    }

    fs::write(PATH, contents).unwrap();
}
